test/e2e/network/traffic_policy.go

```diff
@@ -0,0 +1,453 @@
+		if listErr != nil {
+			return false, listErr
+		}
+
+		numEndpoints := 0
+		for _, slice := range slices.Items {
+			numEndpoints += len(slice.Endpoints)
+		}
+		if len(schedulableNodes) > numEndpoints {
+			framework.Logf("Expected %d endpoints, got %d", len(schedulableNodes), numEndpoints)
+			return false, nil
+		}
+
+		finalSlices = slices.Items
+		return true, nil
+	})
+	framework.ExpectNoError(err, "timed out waiting for EndpointSlices to be ready")
+
+	ginkgo.By("having correct hints set for each endpoint")
+	for _, slice := range finalSlices {
+		for _, ep := range slice.Endpoints {
+			if ep.Zone == nil {
+				framework.Failf("Expected endpoint in %s to have zone: %v", slice.Name, ep)
+			}
+			if topologyHints {
+				if ep.Hints == nil || len(ep.Hints.ForZones) == 0 {
+					framework.Failf("Expected endpoint in %s to have hints: %v", slice.Name, ep)
+				}
+				if len(ep.Hints.ForZones) > 1 {
+					framework.Failf("Expected endpoint in %s to have exactly 1 zone hint, got %d: %v", slice.Name, len(ep.Hints.ForZones), ep)
+				}
+				if *ep.Zone != ep.Hints.ForZones[0].Name {
+					framework.Failf("Expected endpoint in %s to have same zone hint, got %s: %v", slice.Name, *ep.Zone, ep)
+				}
+			} else {
+				if ep.Hints != nil && len(ep.Hints.ForZones) > 0 {
+					framework.Failf("Unexpected hints on endpoint in %s: %v", slice.Name, ep)
+				}
+			}
+
+		}
+	}
+
+	nodesByZone := make(map[string]string)
+	zonesWithNodes := make(map[string][]*v1.Node)
+	for _, node := range schedulableNodes {
+		if zone, ok := node.Labels[v1.LabelTopologyZone]; ok {
+			nodesByZone[node.Name] = zone
+			zonesWithNodes[zone] = append(zonesWithNodes[zone], node)
+		}
+	}
+
+	podList, err := c.CoreV1().Pods(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{})
+	framework.ExpectNoError(err)
+	for _, pod := range podList.Items {
+		if zone, ok := nodesByZone[pod.Spec.NodeName]; ok {
+			tr.zonesByPod[pod.Name] = zone
+			tr.nodesByPod[pod.Name] = pod.Spec.NodeName
+		}
+	}
+
+	// use the largest zone to determine which node's port to use for nodeport endpoint requests
+	largestSeen := 0
+	for zone, nodes := range zonesWithNodes {
+		if len(nodes) > largestSeen {
+			largestSeen = len(nodes)
+			tr.endpointZone = zone
+		}
+	}
+	framework.ExpectNotEqual("", tr.endpointZone, "unable to find a zone with more than one node")
+
+	// always send to a node in the largest zone so balancing will result in more than 1 host name being seen
+	tr.endpointNode = zonesWithNodes[tr.endpointZone][0]
+	tr.peerNode = zonesWithNodes[tr.endpointZone][1]
+
+	// setup the endpoints which are used as target of curl in configurations below
+	nodeIp, err := e2enode.GetInternalIP(tr.endpointNode)
+	framework.ExpectNoError(err, "unable to GetInternalIP for node "+tr.endpointNode.Name)
+
+	// nodeIp includes SSH port, so drop that off
+	nodeIp = strings.SplitN(nodeIp, ":", 2)[0]
```

Comment:
does it? is this coming from another test? NodeIPs are addresses without ports

Reply:
`GetInternalIp` returns the ip joined with ssh port
https://github.com/kubernetes/kubernetes/blob/215f236a6b9087c63f4f65fd34a066e56fae54d7/test/e2e/framework/node/resource.go#L242

Reply:
I see, I didn't know, sorry.

For making the test compatible with IPv4, IPv6 and dual stack we use to do this

```
		family := v1.IPv4Protocol
		if framework.TestContext.ClusterIsIPv6() {
			family = v1.IPv6Protocol
		}

		ips := e2enode.GetAddressesByTypeAndFamily(&nodes.Items[0], v1.NodeInternalIP, family)
		framework.ExpectNotEqual(len(ips), 0)
                nodeIP := ips[0]
```

check if that makes sense or current  approach is enough , please

Reply:
I switched to using your suggested snippet to allow ip6, thanks! Resolving.